src/Core/Grid/Query/ProductQueryBuilder.php

```diff
@@ -0,0 +1,273 @@
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function getSearchQueryBuilder(SearchCriteriaInterface $searchCriteria): QueryBuilder
+    {
+        $qb = $this->getQueryBuilder($searchCriteria->getFilters());
+        $qb
+            ->select('p.`id_product`, p.`reference`')
+            ->addSelect('ps.`price` AS `price_tax_excluded`, ps.`active`')
+            ->addSelect('pl.`name` , pl.`link_rewrite`')
+            ->addSelect('cl.`name` AS `category`')
+            ->addSelect('img_shop.`id_image`')
+        ;
+
+        if ($this->configuration->get('PS_STOCK_MANAGEMENT')) {
+            $qb->addSelect('sa.`quantity`');
+        }
+
+        $this->searchCriteriaApplicator
+            ->applyPagination($searchCriteria, $qb)
+            ->applySorting($searchCriteria, $qb)
+        ;
+
+        return $qb;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function getCountQueryBuilder(SearchCriteriaInterface $searchCriteria): QueryBuilder
+    {
+        $qb = $this->getQueryBuilder($searchCriteria->getFilters());
+        $qb->select('COUNT(p.`id_product`)');
+
+        return $qb;
+    }
+
+    /**
+     * Gets query builder.
+     *
+     * @param array $filterValues
+     *
+     * @return QueryBuilder
+     */
+    private function getQueryBuilder(array $filterValues): QueryBuilder
+    {
+        $availableFilters = [
+            'name',
+            'reference',
+            'category',
+            'active',
+        ];
+
+        $qb = $this->connection
+            ->createQueryBuilder()
+            ->from($this->dbPrefix . 'product', 'p')
+            ->innerJoin(
+                'p',
+                $this->dbPrefix . 'product_shop',
+                'ps',
+                'ps.`id_product` = p.`id_product` AND ps.`id_shop` = :id_shop'
+            )
+            ->leftJoin(
+                'p',
+                $this->dbPrefix . 'product_lang',
+                'pl',
+                'pl.`id_product` = p.`id_product` AND pl.`id_lang` = :id_lang AND pl.`id_shop` = :id_shop'
+            )
+            ->leftJoin(
+                'ps',
+                $this->dbPrefix . 'category_lang',
+                'cl',
+                'cl.`id_category` = ps.`id_category_default` AND cl.`id_lang` = :id_lang AND cl.`id_shop` = :id_shop'
+            )
+            ->leftJoin(
+                'ps',
+                $this->dbPrefix . 'image_shop',
+                'img_shop',
+                'img_shop.`id_product` = ps.`id_product` AND img_shop.`cover` = 1 AND img_shop.`id_shop` = :id_shop'
+            )
+            ->andWhere('p.`state`=1')
+        ;
+
+        $isStockManagementEnabled = $this->configuration->get('PS_STOCK_MANAGEMENT');
+
+        if ($isStockManagementEnabled) {
+            $stockOnCondition =
+                'sa.`id_product` = p.`id_product`
+                    AND sa.`id_product_attribute` = 0
+                ';
+
+            if ($this->isStockSharingBetweenShopGroupEnabled) {
```

Comment:
Woah 😮  so much different usecases.
Can we keep a list of these for QA ? No need to be very explicit, just "when setting A is enabled, we should see B"

Reply:
I have updated how to test section. It's gotten quite big.

I wonder if we need Q&A for this first step? While if there are any issues it would be useful to tackle them as soon as possible Q&A will probably need to test all those things again when I do second part for the product list, to make sure my next PR doesn't break anything from the first PR. 

Also I won't do any tests for first PR, I don't see any behat tests I could write here. Maybe query builder, but do we write tests for those? 
And I thought it would be useful to have tests for filters and stuff like that, but those are already done for product list. Obviously they are not testing v2 route, but we can rely on those once we switch url from v2. (And we will probably need new tests if we implement the "Use new product page" button).

Reply:
Adding Unit tests will be better @matks

Reply:
@Progi1984 Hi ! We had a deeper look to see what we could unit test inside this QueryBuilder but since this QueryBuilder mainly aims to output a configured QueryBuilder ready-to-use it looks like testing what kind of SQL queries we are actually writing.

I have always been told that it was pointless to verify that a piece of php code outputs the right SQL query 🤔  because even if the query is right, it does not mean it will return the right result. Also this kind of test stays green when SQL schema evolves although the query is wrong. Instead I was always told "if your php code interacts closely with database and the logic in it rely on it, use an integration test so that the database logic, which is key to your php code, is tested too".

So far I don't think we wrote integration Behat tests for QueryBuilder but we have a kind of integration tests: the UI tests written by the QA automation team. By testing the sorts, the filters, the searches, they do test the full layer of logic between the view and the database.

Reply:
We discussed in Slack together about integration tests to validate this